filters/filter.go

```diff
@@ -0,0 +1,485 @@
+// Package filters defines a syntax and parser that can be used for the
+// filtration of items across the containerd API. The core is built on the
+// concept of protobuf field paths, with quoting.  Several operators allow the
+// user to flexibly select items based on field presence, equality, inequality
+// and regular expressions. Flexible adaptors support working with any type.
+//
+// The syntax is fairly familiar, if you've used container ecosystem
+// projects.  At the core, we base it on the concept of protobuf field
+// paths, augmenting with the ability to quote portions of the field path
+// to match arbitrary labels. These "selectors" come in the following
+// syntax:
+//
+// ```
+// <fieldpath>[<operator><value>]
+// ```
+//
+// A basic example is as follows:
+//
+// ```
+// name=foo
+// ```
+//
+// This would match all objects that have a field `name` with the value
+// `foo`. If we only want to test if the field is present, we can omit the
+// operator. This is most useful for matching labels in containerd. The
+// following will match objects that have the field "labels" and have the
+// label "foo" defined:
+//
+// ```
+// labels.foo
+// ```
+//
+// We also allow for quoting of parts of the field path to allow matching
+// of arbitrary items:
+//
+// ```
+// labels."very complex label"=something
+// ```
+//
+// We also define `!=` and `:` as operators. The `!=` will match all
+// objects that don't match the value for a field and `:` will compile the
+// target value as an anchored regular expression and match the field value
+// against that.
+//
+// Selectors can be combined using a comma, such that the resulting
+// selector will require all selectors are matched for the object to match.
+// The following example will match objects that are named `foo` and have
+// the label `bar`:
+//
+// ```
+// name=foo,labels.bar
+// ```
+//
+package filters
+
+import (
+	"fmt"
+	"io"
+	"regexp"
+
+	"github.com/containerd/containerd/log"
+	"github.com/pkg/errors"
+)
+
+var (
+	errInvalidFormat   = errors.New("invalid format")
```

Comment:
nit: I actually realized that I don't like defining an error object here since the stacktrace is also captured at that moment. So unless we wrap them again when returning them, this useful information is lost.